server/lib/src/idm/application.rs

```diff
@@ -0,0 +1,715 @@
+use super::ldap::{LdapBoundToken, LdapSession};
+use crate::idm::account::Account;
+use crate::idm::event::LdapApplicationAuthEvent;
+use crate::idm::server::{IdmServerAuthTransaction, IdmServerTransaction};
+use crate::prelude::*;
+use concread::cowcell::*;
+use hashbrown::HashMap;
+use kanidm_proto::internal::OperationError;
+use std::sync::Arc;
+use uuid::Uuid;
+
+#[derive(Clone)]
+pub(crate) struct Application {
+    pub uuid: Uuid,
+    pub name: String,
+    pub linked_group: Option<Uuid>,
+}
+
+impl Application {
+    #[cfg(test)]
+    pub(crate) fn try_from_entry_ro(
+        value: &Entry<EntrySealed, EntryCommitted>,
+        _qs: &mut QueryServerReadTransaction,
+    ) -> Result<Self, OperationError> {
+        if !value.attribute_equality(Attribute::Class, &EntryClass::Application.to_partialvalue()) {
+            return Err(OperationError::InvalidAccountState(
+                "Missing class: application".to_string(),
+            ));
+        }
+
+        let uuid = value.get_uuid();
+
+        let name = value
+            .get_ava_single_iname(Attribute::Name)
+            .map(|s| s.to_string())
+            .ok_or_else(|| {
+                OperationError::InvalidAccountState(format!(
+                    "Missing attribute: {}",
+                    Attribute::Name
+                ))
+            })?;
+
+        let linked_group: Option<Uuid> = value.get_ava_single_refer(Attribute::LinkedGroup);
+
+        Ok(Application {
+            name,
+            uuid,
+            linked_group,
+        })
+    }
+}
+
+#[derive(Clone)]
+struct LdapApplicationsInner {
+    set: HashMap<String, Application>,
```

Comment:
What's the logic of using a string to index into this Map instead of UUID?

Reply:
Mainly because `LdapApplicationAuthEvent` carries the application name extracted from the bind DN, so we can do https://github.com/scabrero/kanidm/blob/app-passwords/server/lib/src/idm/application.rs#L178